### PR TITLE
Add periodic checks with sanitizers

### PR DESCRIPTION
**Pull request by daljit46**
Sanitizers are very valuable tools to detect problems by instrumenting code for runtime bug detection. They have already proven useful for us as demonstrated in #2760, #2761 and #2755. 
This PR adds a new GitHub Actions workflow that runs our unit and binary tests compiled with sanitizers provided by [LLVM](https://github.com/google/sanitizers) each night. It'd be rather computationally expensive to run these checks on pull requests, but my tests show that they are fast enough to be run once a day (with caching enabled).

Note that this PR must be merged into `master` since scheduled GitHub Actions only run on the default or base branch.

### CODE REVIEW

**Comment by daljit46** [comment]
After the discussion today, I think it may make sense to wait for this PR to be merged until most issues spotted by the sanitizers have been fixed on dev.
Also, I think it may make sense to schedule this weekly, instead of daily (although we should try to make sure that CI cache isn't invalidated between consecutive runs).

**Comment by daljit46** [comment]
I think this is mostly ready to be merged. All of issues mentioned in #2772, seems to have fixed apart from #2779. I think that's still waiting a review from @jdtournier.
Once that's merged, I don't see any problem with merging to master.

**Comment by daljit46** [comment]
All tests are now passing, so I think this is now mergeable.

**Review by Lestropie — APPROVED** [review]

**Pull request merged**
